Release v0.2.0

Bullrich opened the pull request:
Base release with the following features

- `get_input`
- `set_output`
- `get_context`: Returns a hydrated context object
- logging
  - `debug`
  - `info`
  - `notice` (with annotations)
  - `warn` (with annotations)
  - `error` (with annotations)
  - `is_debug`

Bullrich committed:
added codeowners

Bullrich committed:
Added context object (#1)

- Created `Context` ([GitHub's documentation for
Contexts](https://docs.github.com/en/actions/learn-github-actions/contexts))
- Added CI tests


## Summary

- **New Features**
- Introduced a continuous integration workflow to automatically run
tests, formatting, linting, and documentation checks on pull requests
and pushes to the main branch.
- Added a `Context` struct to handle context objects injected by GitHub
actions, enabling better retrieval and parsing of event details and
repository information.

- **Dependencies**
  - Added `json` version `0.12.4` as a new dependency.

- **Documentation**
- Included the content of `README.md` in the main module's documentation
for better reference.

Bullrich committed:
added script to publish crate automatically (#2)

Bullrich committed:
fixed CI not working with deprecated package (#3)

Replaced [ciiiii/toml-editor](https://github.com/ciiiii/toml-editor) as
it is not working for a custom pip package.


## Summary


- **Chores**
- Improved the publish workflow by setting up Python and using
`toml-cli` for version management.

Bullrich committed:
added documentation to the Cargo.toml file (#4)

Added all missing information like author, repository and license

Bullrich committed:
Renamed `main.rs` to `lib.rs`

Else it will not work as a library

Bullrich committed:
Fixed issues in `lib` migration (#5)

Fixed some issues I didn't handle when migrating to a library.

- added `allow-dirty` to the library being pushed (as the version is
updated)
- removed unused `main`

Bullrich committed:
improved documentation

Bullrich committed:
added `get_input` method (#6)

- Copied method from
[core.ts](https://github.com/actions/toolkit/blob/d1df13e178816d69d96bdc5c753b36a66ad03728/packages/core/src/core.ts#L126)
- Extended error to handle lack of input method.

## Summary 

- **New Features**
- Introduced `get_input` function to handle input retrieval with
space-to-underscore conversion.
  - Added `InputNotFound` error variant for better error handling.

Bullrich committed:
Added `set_output` method (#8)

- Added method to produce an
[output](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter)
- Copied from
[here](https://github.com/actions/toolkit/blob/d1df13e178816d69d96bdc5c753b36a66ad03728/packages/core/src/core.ts#L192)

## Summary

- **Documentation**
- Updated `README.md` to mark the `set_output` method as complete and
included an example of its usage.

- **Bug Fixes**
- Enhanced error handling with a new `Output` variant in the
`ActionsError` enum.

- **Chores**
  - Added `uuid` dependency with `v4` feature for better UUID handling.

Bullrich committed:
Created logging functions (#9)

Created many methods for logging
- `debug_log`
- `info`
- `warn_log`
- `error_log`
- `notice_log`
- `is_debug`

## Summary

- **New Features**
- Introduced a new logging module with methods for debug, info, warning,
error, and notice messages.
  - Added a function to check if the code is running on a debug runner.

- **Documentation**
- Updated the documentation link in the `Cargo.toml` file and README.md.
- README.md now includes a badge for GitHub Actions and updated code
snippets for usage and installation.

- **Bug Fixes**
- Enhanced error handling for the `set_output` function in the core
module.

Bullrich committed:
removed panic from context::get_env (#10)

Now it returns an error.

Fixes #7

Bullrich committed:
fixed fmt ℹ

Bullrich committed:
Improved CI ⏩  (#11)

- Added benchmarks with the library
[divan](https://crates.io/crates/divan)
- Added more steps to the required tests


## Summary

- **New Features**
- Introduced benchmarking for input/output handling and logging to
improve performance insights.

- **Chores**
  - Added `divan` dependency for benchmarking.
  - Updated build workflow to include new benchmark steps.

Bullrich committed:
added parameters to logs (#12)

Now the users can point to specific failing lines


## Summary

- **New Features**
- Enhanced logging capabilities with new `LogParameters` for more
detailed log information.
  - Environment variable `LOG_DEBUG` introduced for better log control.

Bullrich committed:
removed work in progress from README

It currently has all the MVP features, so we can properly release it as a working library

More work to come tho

Bullrich committed:
📄 updated README to show current available features

Bullrich committed:
updated readme example

mutantcornholio commented on src/context.rs:
```diff
@@ -0,0 +1,132 @@
+    let mut payload: JsonValue = JsonValue::Null;
+
+    if let Ok(github_event_path) = env::var("GITHUB_EVENT_PATH") {
+        if Path::new(&github_event_path).exists() {
+            match fs::read_to_string(&github_event_path) {
+                Ok(content) => match json::parse(&content) {
+                    Ok(parsed_json) => payload = parsed_json,
+                    Err(err) => println!("Failed to parse JSON: {}", err),
+                },
+                Err(err) => println!("Failed to read file: {}", err),
+            }
+        } else {
+            println!("GITHUB_EVENT_PATH {} does not exist", github_event_path)
+        }
```
Checking that file exists before reading is a discouraged practice. Even though it's fine in most cases, there aren't any guarantees that file will persist between check and the read. 

Instead, attempt reading file, and handle the "not found" error as you do here. 
Something like this:
```rust
            match fs::read_to_string(&github_event_path) {
                Ok(content) => match json::parse(&content) {
                    Ok(parsed_json) => payload = parsed_json,
                    Err(err) => println!("Failed to parse JSON: {}", err),
                },
                Err(err) if err.kind() === ErrorKind::NotFound => {
                    println!("GITHUB_EVENT_PATH {} does not exist", github_event_path),
                },
                Err(err) => println!("Failed to read file: {}", err),
            }
```
However, with two error handlers being basically identical, you can simply drop the `if Path::new(&github_event_path).exists() { ` clause.

Bullrich committed:
Optimized context method (#15)

Thanks to @mutantcornholio in #13, we can optimize the method to read
the context to follow better practices.


## Summary

- **Refactor**
- Improved error handling for reading and parsing JSON files to enhance
reliability and user experience.

Bullrich commented:
Has already been merged.
